pkg/informer/informer.go

```diff
@@ -208,45 +202,46 @@ func (d *DynamicDiscoverySharedInformerFactory) discoverTypes(ctx context.Contex
 					continue
 				}
 				if !sets.NewString([]string(ai.Verbs)...).HasAll("list", "watch") {
-					klog.V(4).InfoS("resource is not list+watchable", "logical-cluster", logicalClusterName, "group", gv.Group, "version", gv.Version, "resource", ai.Name, "verbs", ai.Verbs)
 					continue
 				}
 
-				latest.Insert(strings.Join([]string{ai.Name, gv.Version, gv.Group}, "."))
+				logicalClusterNames[gvr] = logicalClusterName
+				latest[gvr] = struct{}{}
 			}
 		}
 	}
 
+	d.mu.RLock()
+	var newGVRs []schema.GroupVersionResource
+	for gvr := range latest {
+		if _, found := d.gvrs[gvr]; !found {
+			newGVRs = append(newGVRs, gvr)
+		}
+	}
+	d.mu.RUnlock()
+
+	if len(newGVRs) == 0 {
+		return nil
+	}
+
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
 	// Set up informers for any new types that have been discovered.
 	// TODO: Stop informers for types we no longer have.
-	newGVRs := latest.Difference(d.gvrs)
-	var merr error
-	for _, gvrstr := range newGVRs.UnsortedList() {
-		klog.Infof("Adding informer for %q in %s", gvrstr, logicalClusterNames)
-		gvr, _ := schema.ParseResourceArg(gvrstr)
-		if gvr == nil {
-			// TODO(ncdc): consider tracking where each gvrstr came from (which workspace) so we can include that in the error.
-			multierr.AppendInto(&merr, fmt.Errorf("unable to parse %q to a GroupVersionResource", gvrstr))
-			continue
-		}
-
-		inf := d.dsif.ForResource(*gvr).Informer()
+	for _, gvr := range newGVRs {
+		klog.Infof("Adding informer for %q in workspace %s", gvr, logicalClusterNames)
```

Comment:
IIRC these are wildcard informers, so we can probably remove the part about which workspace the gvr came from, unless you want to log that it was this workspace that triggered the add?

Reply:
Yes, that's the intention to understand what triggered the addition. Also note that we have no logic to remove them when the resource goes away.